### PR TITLE
[GEARPUMP-252] return meaningful result than app id when submitting app…

### PR DESCRIPTION
**Pull request by huafengw**
…lication in ClientContext

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

**Comment by huafengw** [comment]
@manuzhang

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      case Success(_) =>
+      case Failure(ex) => throw ex
+    }
+  }
+
+  def askMaster[T](msg: Any): Future[T] = {
+    appMaster.ask(msg)(timeout).asInstanceOf[Future[T]]
+  }
+
+//  Todo
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: I'm afraid the intention for these lines could be forgotten after a while so maybe better just remove them for now.

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import akka.util.Timeout
+import org.apache.gearpump.cluster.ClientToMaster.{ResolveAppId, ShutdownApplication}
+import org.apache.gearpump.cluster.MasterToClient.{ResolveAppIdResult, ShutdownApplicationResult}
+import org.apache.gearpump.util.ActorUtil
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
+  import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: better put this in the header

File: core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala [hunk 3]
```diff
@@ -75,59 +78,67 @@ class ClientContext(config: Config, sys: ActorSystem, _master: ActorRef) {
...
 
   private def getSubmissionConfig(config: Config): Config = {
     ClusterConfig.filterOutDefaultConfig(config)
   }
 
+  def listApps: AppMastersData = {
+    ActorUtil.askActor[AppMastersData](master, AppMastersDataRequest, masterClientTimeout)
+  }
+
+  @Deprecated
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala]: If these methods are deprecated, which should be used instead ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import akka.actor.ActorRef
+import akka.util.Timeout
+import org.apache.gearpump.cluster.ClientToMaster.{ResolveAppId, ShutdownApplication}
+import org.apache.gearpump.cluster.MasterToClient.{ResolveAppIdResult, ShutdownApplicationResult}
+import org.apache.gearpump.util.ActorUtil
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+class RunningApplication(val appId: Int, master: ActorRef, timeout: Timeout) {
```
Comment (manuzhang) [inline core/src/main/scala/org/apache/gearpump/cluster/client/RunningApplication.scala]: Any UT for this class ?

File: core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala [hunk 5]
```diff
@@ -53,20 +55,21 @@ class ClientContext(config: Config, sys: ActorSystem, _master: ActorRef) {
   }
 
   private val LOG: Logger = LogUtil.getLogger(getClass)
-  private implicit val timeout = Timeout(5, TimeUnit.SECONDS)
-
   implicit val system = Option(sys).getOrElse(ActorSystem(s"client${Util.randInt()}", config))
   LOG.info(s"Starting system ${system.name}")
-  val shouldCleanupSystem = Option(sys).isEmpty
-
+  private val shouldCleanupSystem = Option(sys).isEmpty
```
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala]: The `system` won't be empty but `sys` can still be empty. `sys` not being empty means user has provided a ActorSystem so clientcontext should clean up user's actor system when being closed.

File: core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala [hunk 6]
```diff
@@ -53,20 +55,21 @@ class ClientContext(config: Config, sys: ActorSystem, _master: ActorRef) {
...
   private val jarStoreClient = new JarStoreClient(config, system)
+  private val masterClientTimeout = {
+    val timeout = Try(config.getInt(Constants.GEARPUMP_MASTERCLIENT_TIMEOUT)).getOrElse(90)
```
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/client/ClientContext.scala]: It's inherited from the old impl

**Comment by manuzhang** [comment]
+1, merging

**Pull request closed without merging**